File: solutions/python/perfect-numbers/1/perfect_numbers.py
def classify(number):
    """ A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """

    if number < 1:
        raise ValueError("Classification is only possible for positive integers.")

    aliquot_sum = sum(filter(lambda n: number % n == 0, range(1, number)))

    if aliquot_sum == number:
        return "perfect"

    if aliquot_sum > number:
        return "abundant"

    return "deficient"

File: solutions/csharp/perfect-numbers/1/PerfectNumbers.cs
using System;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        if (number < 1)
            throw new ArgumentOutOfRangeException();

        // edge case
        if (number == 1)
            return Classification.Deficient;

        var root = (int)Math.Sqrt(number);

        var aliquot_sum = 1; // we're starting with 2, but 1 is a valid factor
        for (var factor = 2; factor <= root; ++factor)
        {
            if (number % factor == 0)
            {
                aliquot_sum += factor;

                var complementary = number / factor;
                if (complementary != factor)
                    aliquot_sum += complementary;
            }
        }

        if (aliquot_sum == number)
            return Classification.Perfect;

        if (aliquot_sum > number)
            return Classification.Abundant;

        return Classification.Deficient;
    }
}
